### PR TITLE
Apfelbaeumle

### PR DESCRIPTION
**Pull request by apfelbaeumle**
* implement update for groups, group_membership in server
* change listen_users so that updates also send an update of the users a user is allowed to see as they are members of one of his groups
* implement some typechecks
* add crpc to request groups and users by ids. The server of course only sends them if the user is allowed to see them
* client: add a structure (cache) to store all actual valid data it got from the server
* client: implement listen_groups, listen_group_memberships (they simply store the updated into the cache and log it to the console, no visualisation yet)
* remove some dead code in server/main.js

### CODE REVIEW

File: server/eval/typecheck.js [hunk 1]
```diff
@@ -15,12 +15,18 @@ module.exports.object = (object) => ((typeof object === "object") && !Array.isAr
 module.exports.bool = (object) => (object === true) || (object === false);
 module.exports.number = Number.isFinite;
 module.exports.number_nonnegative = (object) => (Number.isFinite(object) && (object >= 0));
+module.exports.integer = Number.isInteger;
+module.exports.positive = (object) => (object > 0);
```
Comment (mic-e) [inline server/eval/typecheck.js]: add a comment that this can only be used in combination with other typechecks

File: server/main.js [hunk 2]
```diff
@@ -305,112 +350,201 @@ crpc_functions.listen_groups = async (connection, args) => {
...
-            delete update.json;
-            delete update.comment;
-        }
-        connection.sendUTF(JSON.stringify(update));
+crpc_functions.get_groups_by_id = async (connection, args) => {
+    typecheck.validate_object_structure(args, ARGS_GET_GROUPS_BY_ID);
+
+    const self_uid = await connection.get_user();
+
+    const ids = args.ids.filter( () => true ); 		// make array dense: remove any elem which is unset
```
Comment (mic-e) [inline server/main.js]: please add comment in the line before

File: server/main.js [hunk 3]
```diff
@@ -305,112 +350,201 @@ crpc_functions.listen_groups = async (connection, args) => {
...
+                group_memberships
+            where
+                groups.id = group_memberships.gid and
+                group_memberships.uid = $1 and
+                groups.id in (${callparams})
+        ), max_last_mod_seq as (
+            select max(last_mod_seq) as max_last_mod_seq from groups_of_user
+        )
+        select
+            id,
```
Comment (mic-e) [inline server/main.js:481]: please add a comment that these fields are redundant with the ones in `listen_groups`

File: server/main.js [hunk 4]
```diff
@@ -305,112 +350,201 @@ crpc_functions.listen_groups = async (connection, args) => {
...
+        return { "users": [] };
     }
-    return { idx: update.idx, event: update.event };
-}
-*/
+
+    let paramnr = 2;
+    const callparams = ids.map( () => "$"+(paramnr++) ).join(",");
+
+    const users = (await connection.db.query(
```
Comment (mic-e) [inline server/main.js:517]: please add a comment regarding redundancy

File: server/eval/typecheck.js (hunk 1 above)
Comment (apfelbaeumle) [reply server/eval/typecheck.js]: done

File: server/main.js (hunk 3 above)
Comment (apfelbaeumle) [reply server/main.js:481]: done

File: server/main.js (hunk 4 above)
Comment (apfelbaeumle) [reply server/main.js:517]: done

**Pull request merged**